block/manager.go

```diff
@@ -212,13 +217,25 @@ func (m *Manager) healthStatusEventCallback(event pubsub.Message) {
 }
 
 func (m *Manager) applyBlockCallback(event pubsub.Message) {
-	m.logger.Debug("Received new block event", "eventData", event.Data())
+	m.logger.Debug("Received new block event", "eventData", event.Data(), "cachedBlocks", len(m.prevBlock))
 	eventData := event.Data().(p2p.GossipedBlock)
 	block := eventData.Block
 	commit := eventData.Commit
-	err := m.applyBlock(context.Background(), &block, &commit, blockMetaData{source: gossipedBlock})
-	if err != nil {
-		m.logger.Debug("Failed to apply block", "err", err)
+
+	if block.Header.Height != m.store.Height()+1 {
+		// We crashed after the commit and before updating the store height.
+		m.prevBlock[block.Header.Height] = &block
```

Comment:
We should probably try to apply the cached blocks even if we just applied a block because and not just if the blocks are out of order. for example,  if we got previously block at (store height + 2) and put it into cache  and now we got storeHeight + 1 (which means we won't go into the "if")  we can still apply the cached block and no need to wait for the next out-of-order block in order to try it.. 

I would simply add at the end of the function a check, if the block we have in cache is == m.storeHeight + 1 then run `ApplyCachedBlock`.

Reply:
true. having function call in the if condition may need an extra block to apply all cached blocks. it's been changed